util/types/json/compare.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package json
+
+import (
+	"bytes"
+	"reflect"
+	"strings"
+)
+
+// floatEpsilon is for compare two float value allowing precision loss.
+const floatEpsilon float64 = 0.00000001
+
+// compareFloat64 returns an integer comparing the float64 x to y,
+// allowing precision loss.
+func compareFloat64PrecisionLoss(x, y float64) int {
+	if x-y < floatEpsilon && y-x < floatEpsilon {
+		return 0
+	} else if x-y < 0 {
+		return -1
+	}
+	return 1
+}
+
+// jsonTypePrecedences is for compare two json.
+// See: https://dev.mysql.com/doc/refman/5.7/en/json.html#json-comparison
+var jsonTypePrecedences = map[string]int{
+	"BLOB":     -1,
+	"BIT":      -2,
+	"OPAQUE":   -3,
+	"DATETIME": -4,
+	"TIME":     -5,
+	"DATE":     -6,
+	"BOOLEAN":  -7,
+	"ARRAY":    -8,
+	"OBJECT":   -9,
+	"STRING":   -10,
+	"INTEGER":  -11,
+	"DOUBLE":   -11,
+	"NULL":     -12,
+}
+
+func jsonAsFloat64(j JSON) float64 {
+	if j.Type() == "INTEGER" {
+		if reflect.TypeOf(j).Kind() == reflect.Ptr {
+			return float64(*j.(*jsonInt64))
+		}
+		return float64(j.(jsonInt64))
+	} else if j.Type() == "DOUBLE" {
+		if reflect.TypeOf(j).Kind() == reflect.Ptr {
+			return float64(*j.(*jsonDouble))
+		}
+		return float64(j.(jsonDouble))
+	} else {
+		panic("can not convert to float64")
+	}
+}
+
+func jsonAsString(j JSON) string {
+	if reflect.TypeOf(j).Kind() == reflect.Ptr {
+		return string(*j.(*jsonString))
+	}
+	return string(j.(jsonString))
+}
+
+// CompareJSON compares two json object.
+func CompareJSON(j1 JSON, j2 JSON) (cmp int, err error) {
+	precedence1 := jsonTypePrecedences[j1.Type()]
+	precedence2 := jsonTypePrecedences[j2.Type()]
+
+	if precedence1 == precedence2 {
+		if precedence1 == -12 {
```

Comment:
use jsonTypePrecedences["NULL"] here may be more explicit.
the same as line 121 and line 125